budgetportal/prov_infra_project/charts.py

```diff
@@ -0,0 +1,135 @@
+import simplejson as json
+
+
+def order_chart_data(snapshot_list):
+    return sorted(snapshot_list, key=lambda x: x["date"])
+
+
+def time_series_data(project):
+    chart_data = {"snapshots": [], "events": []}
+    project_snapshots = project.project_snapshots.all()
+    for snapshot in project_snapshots:
+        quarter_number = snapshot.irm_snapshot.quarter.number
+        fin_year = snapshot.irm_snapshot.financial_year
+
+        total_estimated_project_cost = snapshot.total_project_cost
+        status = snapshot.status
+        date, quarter_label, fin_year_label = extract_date_quarter_year(
+            quarter_number, fin_year
+        )
+        total_spent_to_date, total_spent_in_quarter = compute_total_spent(
+            snapshot, quarter_number
+        )
+        chart_data["snapshots"].append(
+            {
+                "date": date,
+                "quarter_label": quarter_label,
+                "financial_year_label": fin_year_label,
+                "total_spent_to_date": total_spent_to_date,
+                "total_spent_in_quarter": total_spent_in_quarter,
+                "total_estimated_project_cost": total_estimated_project_cost,
+                "status": status,
+            }
+        )
+        if quarter_number > 1:
+            chart_data = update_previous_chart_values(
+                chart_data, snapshot, quarter_number, fin_year
+            )
+    chart_data["snapshots"] = order_chart_data(chart_data["snapshots"])
+
+    latest_snapshot = find_latest_snapshot_with_dates(project)
+    if latest_snapshot:
+        chart_data["events"] = extract_events(latest_snapshot)
+
+    return json.dumps(chart_data, use_decimal=True)
```

Comment:
this should just return the chart_data object - the caller will encode as json. But the output of this should be nested dictionaries and lists that are compativle with json.dumps. So it looks like we should have a test that verifies that the output of this function can be passed to `json.dumps`

Why are you using simplejson? The task was to use `json.dumps` - by that I meant the json package built into python.

I'm using a customised encoder to format dates in `budgetportal.webflow.views`

Reply:
Perhaps it's a good time to move the encoder to a module like `budgetportal.json` then it can be used in this test and in the webflow views.

Reply:
Encoder moved.

https://github.com/vulekamali/datamanager/pull/215/commits/f4a9dc067dbf8387d215c359592855f043b29d4d